sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogBeamSchema.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.hcatalog;
+
+import com.sun.istack.Nullable;
+import java.util.Map;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.base.Optional;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+
+/**
+ * Adapter from HCatalog table schema to Beam {@link org.apache.beam.sdk.schemas.Schema}.
+ *
+ * <p>Loads a table schema from Hive Metastore specified in properties map, similar to {@link
+ * HCatalogIO}.
+ *
+ * <p>One of the use cases is to perform the schema conversion without leaking any HCatalog types.
+ */
+@Experimental
+public class HCatalogBeamSchema {
+
+  private @Nullable IMetaStoreClient metastore;
+
+  private HCatalogBeamSchema(IMetaStoreClient metastore) {
+    this.metastore = metastore;
+  }
+
+  /**
+   * Create the schema adapter.
+   *
+   * <p>Config map is used to construct the {@link HiveMetaStoreClient}.
+   */
+  public static HCatalogBeamSchema create(Map<String, String> config) {
+    try {
+      HiveConf hiveConf = new HiveConf();
+      config.forEach(hiveConf::set);
+      return new HCatalogBeamSchema(new HiveMetaStoreClient(hiveConf));
+    } catch (Exception e) {
+      throw new RuntimeException(e);
```

Comment:
A good practice to connect hive metastore is retry somewhere (either in hive client or here) when having some exceptions, which should reduce the failure ratio of this code path. However, currently this class is experimental so we can wait and see if there is any report from users on failure ratio.

Reply:
I don't have enough knowledge of Hive to add a meaningful retry policy. However, it seems that retrying already [works to some degree](https://github.com/apache/hive/blob/35f86c749cefc2a9972a991deed78a1c3719093d/standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/RetryingMetaStoreClient.java#L52), if users [do not specify the FASTPATH config option](https://github.com/apache/hive/blob/71dfd1d11f239caf8f16bc29db0f959e566f7659/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java#L4928).

Reply:
My understanding to that retry is that is used for retry connections (a connection breaks then it tries to reconnect or fail to connect and reconnect). Sometime, in a normal connection, metastore could return a high level recoverable exception(e.g. possible database exception), which we might can retry issue command to metastore.  I cannot recall the details of those high level exceptions though.